src/Coordination/KeeperDispatcher.cpp

```diff
@@ -610,6 +769,78 @@ void KeeperDispatcher::updateConfigurationThread()
+
+    for (const auto & request_for_session : requests_for_sessions)
+    {
+        ++counts_for_session[request_for_session.session_id];
+    }
+
+    std::lock_guard lock{unprocessed_request_mutex};
+    for (const auto [session_id, count] : counts_for_session)
+    {
+        auto unprocessed_requests_it = unprocessed_requests_for_session.find(session_id);
+        if (unprocessed_requests_it == unprocessed_requests_for_session.end())
+            continue;
+
+        auto & unprocessed_requests = unprocessed_requests_it->second;
+        unprocessed_requests.unprocessed_num -= count;
+
+        if (unprocessed_requests.unprocessed_num == 0)
+        {
+            if (!unprocessed_requests.request_queue.empty())
+            {
+                auto & request_queue = unprocessed_requests.request_queue;
+                unprocessed_requests.is_read = !unprocessed_requests.is_read;
+                // start adding next type of requests
+                while (!request_queue.empty() && request_queue.front().request->isReadRequest() == unprocessed_requests.is_read)
+                {
+                    auto & front_request = request_queue.front();
+
+                    /// Put close requests without timeouts
+                    if (front_request.request->getOpNum() == Coordination::OpNum::Close)
+                    {
+                        if (!requests_queue->push(std::move(front_request)))
+                            throw Exception("Cannot push request to queue", ErrorCodes::SYSTEM_ERROR);
+                    }
+                    else if (!requests_queue->tryPush(std::move(front_request), configuration_and_settings->coordination_settings->operation_timeout_ms.totalMilliseconds()))
+                    {
+                        throw Exception("Cannot push request to queue within operation timeout", ErrorCodes::TIMEOUT_EXCEEDED);
+                    }
+
+                    ++unprocessed_requests.unprocessed_num;
+                    request_queue.pop_front();
+                }
+            }
+            else
+            {
+                unprocessed_requests_for_session.erase(unprocessed_requests_it);
+            }
+        }
+    }
+}
+
+void KeeperDispatcher::onRequestCommit(const KeeperStorage::RequestForSession & request_for_session, uint64_t log_term, uint64_t log_idx)
+{
+    if (!finalize_requests_queue.push({request_for_session}))
+        throw Exception(ErrorCodes::SYSTEM_ERROR, "Cannot push read requests to queue");
+
+    KeeperStorage::RequestsForSessions requests;
+    {
+        std::lock_guard lock(leader_waiter_mutex);
+        auto request_queue_it = leader_waiters.find(KeeperServer::NodeInfo{.term = log_term, .last_committed_index = log_idx});
+        if (request_queue_it != leader_waiters.end())
+        {
+            requests = std::move(request_queue_it->second);
+            leader_waiters.erase(request_queue_it);
+        }
+    }
+    if (!read_requests_queue.push(std::move(requests)))
```

Comment:
Cannot understand how we synchronize it and preserve the order of reads and writes with main requests thread...

Reply:
we control per session which types of requests are being currently processed (`unprocessed_requests`)
In all parts of code we make an assumption that at no point both read and write requests are being processed from a single session.

There are no ordering requirements for requests from different sessions.